### PR TITLE
chore(ci): Run all tests on the oldest golang version

### PR DESCRIPTION
**Pull request by machine424**
Run test_mixins on the latest golang version

 
 From https://github.com/prometheus/prometheus/pull/13389#pullrequestreview-1822015642

### CODE REVIEW

**Comment by machine424** [comment]
cc @bboreham as discussed.

**Comment by SuperQ** [comment]
We need to make sure that the `go fmt` test does not run in both versions. We've had issues where we have two incompatible formatting versions in the past.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -8,10 +8,13 @@ jobs:
...
-    # Whenever the Go version is updated here, .promu.yml
-    # should also be updated.
+    strategy:
+      matrix:
+        # Whenever the Go versions are updated here, .promu.yml
+        # should also be updated.
+        go: ["1.20", "1.21"]
     container:
-      image: quay.io/prometheus/golang-builder:1.21-base
+      image: quay.io/prometheus/golang-builder:${{ matrix.go }}-base
```
[the remark this answers is not in the document]
Comment (bboreham) [reply .github/workflows/ci.yml:17]: I don't like that it runs `npm` twice.  Although they run in parallel, I assume this is chewing up some limited resource.

[the remark this answers is not in the document]
Comment (machine424) [reply .github/workflows/ci.yml:17]: I though testing a complete build may make sense. Especially for the part where the UI is embedded. But I don't think we test that after all.
how about running `common-build`? Or should we just go for tests?

[the remark this answers is not in the document]
Comment (machine424) [reply .github/workflows/ci.yml:17]: By the way, the current test runs a complete build: https://github.com/prometheus/prometheus/blob/34875ae8c76b0f1046676050c14cab79d74c06f5/.github/workflows/ci.yml#L72

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -26,6 +26,21 @@ jobs:
         with:
           version: "3.15.8"
 
+  test_go_oldest:
+    name: Go tests with previous Go version
+    runs-on: ubuntu-latest
+    strategy:
```
Comment (machine424) [inline .github/workflows/ci.yml:32]: Remove: the CI was green though 🙃

**Comment by beorn7** [comment]
@bboreham does this look good to you now?

@machine424 assuming @bboreham's comments are addressed, merge conflicts need to get resolved, and then we can merge.

**Comment by machine424** [comment]
Thanks, done @beorn7

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -27,6 +27,20 @@ jobs:
...
+    container:
+      # The go version in this image should be N-1 wrt test_go.
+      image: quay.io/prometheus/golang-builder:1.21-base
+    steps:
+      - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+      - run: make build
+      - run: make test
+      - run: go test ./tsdb/ -test.tsdb-isolation=false
+      - run: go test --tags=stringlabels ./...
+      - run: GOARCH=386 go test ./cmd/prometheus
```
Comment (bboreham) [inline .github/workflows/ci.yml:42]: This seems to be doing 3 complete sets of tests plus one subset.
That feels like overkill to me; the point is to catch code that has been added which is incompatible with the older Go.  Do you think that further incompatibilities will be caught
under different tags or architecture, such that it's worth spending the time?

ISTM that `make test` would be simplest, but I could also argue that `--tags=stringlabels` is the mainline version Prometheus releases with.

Comment (machine424) [reply .github/workflows/ci.yml:42]: If we don't want to take any risks, I'd say yes, we should run the exact go tests than the current version.

- I'd even add the new `go test --tags=dedupelabels ./...`
- I can set  the new GO_ONLY=1 to `make test` to only run go tests.

If we're worried about resources, we can make it a daily test, but that may require more human resources...

Comment (machine424) [reply .github/workflows/ci.yml:42]: We can go with `make test` now, and enrich if we have a slip out in the future.

Comment (bboreham) [reply .github/workflows/ci.yml:42]: The most precious resource is developer time.  If this test is the slowest part, then that's what everyone has to wait for: 
image

I think 22 minutes is way overkill; please just run one set of tests.
"don't want to take any risks" is not something anyone said about Prometheus.

Comment (machine424) [reply .github/workflows/ci.yml]: Let's go with it then.
(Tell me if I should add the test with `-test.tsdb-isolation=false` back)

**Comment by bboreham** [comment]
* Note also #13715

**Pull request merged**